### PR TITLE
Merge modification for Windows support

### PR DESCRIPTION
**Pull request by HerveAncher**
Support Windows compilation with MSVC (Visual Studio 2022) + CMake + vcpkg

### CODE REVIEW

**Comment by HerveAncher** [comment]
Also check that there is no issue with original sources under Linux (CentOS 7).

**Comment by HerveAncher** [comment]
Dear @aka-rider 
Could you please review this PR? The goal is to integrate this library to vcpkg, and allow projects to deploy it as very standard ones.
Regards.

**Comment by aka-rider** [comment]
Hey @HerveAncher 

Thanks a ton for the PR. 
I wanted Windows support for a long time but never got to it.
I will review it ASAP, but it would take a few days as I don't have a development-ready Windows machine or VM.

Thanks again.

Cheers,
Iurii

**Comment by aka-rider** [comment]
@HerveAncher I started to look at it.

Could you please revert changes of `not` keyword to  `!` ? 

I'm not strictly against these changes, but they make PR very hard to review. Moreover, `not` is a valid C++ keyword.

Some more context:

> C++ specifies not as an alternative spelling for !. In C, the alternative spelling is provided as a macro in the <iso646.h> header. In C++, the alternative spelling is a keyword; use of <iso646.h> or the C++ equivalent <ciso646> is deprecated. In Microsoft C++, the [/permissive-](https://learn.microsoft.com/en-us/cpp/build/reference/permissive-standards-conformance?view=msvc-170) or [/Za](https://learn.microsoft.com/en-us/cpp/build/reference/za-ze-disable-language-extensions?view=msvc-170) compiler option is required to enable the alternative spelling.

It would be great if you could enable `/permissive-` compiler option or some of the options that toggle `/permissive-` on implicitly .

> The /permissive- option is implicitly set by the /std:c++latest option starting in Visual Studio 2019 version 16.8, and in version 16.11 by the /std:c++20

**Comment by HerveAncher** [comment]
OK, I will try to introduce `/permissive-` flag and compile with it.

Aside, the PR#33 will be a good introduction for CMake deployment.

**Comment by HerveAncher** [comment]
@aka-rider I just introduced the requested flag, and it works!!
Thanks for this information, I really though that these keywords could not be used under Windows ;)

**Comment by HerveAncher** [comment]
@aka-rider I review the PR#33 of @pavanbadugu and I introduced the modification in this PR.
So normally this PR should cover Cachelot installation and CMake package

**Comment by aka-rider** [comment]
Thank you for making the changes. I wish to react faster, but I'm busy with other things.

I will travel during the next week and will come back to the PR afterward. 
I will take a closer look at the UB checker (I suspect `unaligned_bytes`). Would be great if you could fix it.

File: src/cachelot/bits.h [hunk 1]
```diff
@@ -122,7 +131,7 @@ namespace cachelot {
 
     /// return number of bytes necessary to align `size` according to the given `alignment`
     constexpr size_t unaligned_bytes(const size_t size, const size_t alignment) noexcept {
-        return ((size + (alignment - 1u)) & -alignment) - size;
+        return ((size + (alignment - 1u)) & -((long long)alignment)) - size;
```
[the remark this answers is not in the document]
Comment (HerveAncher) [reply src/cachelot/bits.h:134]: OK, bad cast from my side.
I will replace the `long long` by `intptr_t` which depend of the architecture, better.

File: src/cachelot/string_conv.h [hunk 2]
```diff
@@ -131,7 +131,7 @@ namespace cachelot {
                 UnsignedIntType xvalue;
                 if (negative) {
                     *dest++ = '-';
-                    xvalue = - static_cast<UnsignedIntType>(value);
+                    xvalue = - (long long)static_cast<UnsignedIntType>(value);
```
[the remark this answers is not in the document]
Comment (HerveAncher) [reply src/cachelot/string_conv.h:134]: Same as previous conversation

File: src/cachelot/string_conv.h [hunk 3]
```diff
@@ -331,7 +331,7 @@ namespace cachelot {
                 return unsigned_value * sign;
             }
             if (unsigned_value == int64_boundary && sign == -1) {
-                return -unsigned_value;
+                return -(long long)unsigned_value;
```
[the remark this answers is not in the document]
Comment (HerveAncher) [reply src/cachelot/string_conv.h:334]: Same as previous conversation

File: src/server/memcached/proto_defs.h [hunk 4]
```diff
@@ -21,7 +21,7 @@ namespace cachelot { namespace memcached {
         x(APPEND)               \
         x(CAS)                  \
         x(DECR)                 \
-        x(DELETE)               \
+        x(DEL)                  \
```
[the remark this answers is not in the document]
Comment (HerveAncher) [reply src/server/memcached/proto_defs.h:24]: Yes, `DELETE` is a defined keyword for Windows in winnt.h:
```
#define DELETE (0x00010000L): 
```

I can do like I did with `ERROR` flag, undefine the Windows flag and use `DELETE` in Cachelot.

**Comment by HerveAncher** [comment]
> Thank you for making the changes. I wish to react faster, but I'm busy with other things.
> 
> I will travel during the next week and will come back to the PR afterward. I will take a closer look at the UB checker (I suspect `unaligned_bytes`). Would be great if you could fix it.

About `unaligned_bytes` I made modification for Windows compatibility.
The `struct meta` could not contain something else as integer, so I changed the bool type by uint32, but the alignment works as the same.
And I also forced the packing of the structure, because without this force, the alignment is not good.

**Comment by HerveAncher** [comment]
@aka-rider I think it will be also good, when this PR will be OK, to add tag `1.0.0` for this version.
Currently, when CMake is generating compilation scripts, and during execution bad version is found:

```
./test_c_api
Testing ver. Cachelot v?.?.?-XXX (unknown) Linux x64 C API ....
memory_limit = 4294967296
mem_page_size = 1024
initial_dict_size = 2048
enable_evictions = 1

All tests passes
```

**Comment by HerveAncher** [comment]
I realized that I never give you somme feedback about the perf under Windows.
With the last commit [708bb2d](https://github.com/cachelot/cachelot/pull/37/commits/708bb2da3cc3024da417e77a27ec30df08292432) in Release configuration, Windows 10 x64, i7-12700H + 32 GB of RAM:
```
benchmark_memalloc.exe

 Run #01         498931 allocations
runtime builtin..........    alloc:   190.32 ms     free:   145.12 ms    total:   335.44 ms  (x1.000)
cachelot.................    alloc:    29.95 ms     free:    52.18 ms    total:    82.13 ms  (x4.084)
TLSF.....................    alloc:   104.09 ms     free:    70.99 ms    total:   175.08 ms  (x1.916)


 Run #02         499282 allocations
runtime builtin..........    alloc:   199.19 ms     free:   165.85 ms    total:   365.03 ms  (x1.000)
cachelot.................    alloc:    29.72 ms     free:    53.59 ms    total:    83.31 ms  (x4.382)
TLSF.....................    alloc:    99.39 ms     free:    70.71 ms    total:   170.10 ms  (x2.146)


 Run #03         499527 allocations
runtime builtin..........    alloc:   201.42 ms     free:   167.51 ms    total:   368.93 ms  (x1.000)
cachelot.................    alloc:    29.60 ms     free:    52.71 ms    total:    82.32 ms  (x4.482)
TLSF.....................    alloc:   101.73 ms     free:    71.95 ms    total:   173.69 ms  (x2.124)


 Run #04         500279 allocations
runtime builtin..........    alloc:   203.87 ms     free:   178.83 ms    total:   382.70 ms  (x1.000)
cachelot.................    alloc:    29.60 ms     free:    51.65 ms    total:    81.25 ms  (x4.710)
TLSF.....................    alloc:    95.05 ms     free:    73.68 ms    total:   168.73 ms  (x2.268)


 Run #05         499507 allocations
runtime builtin..........    alloc:   216.31 ms     free:   175.17 ms    total:   391.48 ms  (x1.000)
cachelot.................    alloc:    29.52 ms     free:    52.31 ms    total:    81.83 ms  (x4.784)
TLSF.....................    alloc:    94.48 ms     free:    76.33 ms    total:   170.81 ms  (x2.292)


 Run #06         499930 allocations
runtime builtin..........    alloc:   211.95 ms     free:   174.25 ms    total:   386.20 ms  (x1.000)
cachelot.................    alloc:    29.65 ms     free:    59.33 ms    total:    88.98 ms  (x4.340)
TLSF.....................    alloc:    96.19 ms     free:    77.19 ms    total:   173.37 ms  (x2.228)


 Run #07         498891 allocations
runtime builtin..........    alloc:   202.50 ms     free:   181.33 ms    total:   383.83 ms  (x1.000)
cachelot.................    alloc:    28.87 ms     free:    51.97 ms    total:    80.84 ms  (x4.748)
TLSF.....................    alloc:    94.15 ms     free:    71.86 ms    total:   166.01 ms  (x2.312)


 Run #08         499983 allocations
runtime builtin..........    alloc:   201.90 ms     free:   190.95 ms    total:   392.85 ms  (x1.000)
cachelot.................    alloc:    30.22 ms     free:    51.64 ms    total:    81.86 ms  (x4.799)
TLSF.....................    alloc:    94.63 ms     free:    71.65 ms    total:   166.28 ms  (x2.363)


 Run #09         499169 allocations
runtime builtin..........    alloc:   207.04 ms     free:   180.11 ms    total:   387.15 ms  (x1.000)
cachelot.................    alloc:    28.37 ms     free:    53.82 ms    total:    82.18 ms  (x4.711)
TLSF.....................    alloc:    95.39 ms     free:    73.48 ms    total:   168.87 ms  (x2.293)


 Run #10         499744 allocations
runtime builtin..........    alloc:   199.54 ms     free:   177.62 ms    total:   377.16 ms  (x1.000)
cachelot.................    alloc:    30.41 ms     free:    51.80 ms    total:    82.21 ms  (x4.588)
TLSF.....................    alloc:    94.66 ms     free:   119.21 ms    total:   213.87 ms  (x1.763)
```

**Comment by aka-rider** [comment]
Thanks a lot for your contribution @HerveAncher

Apologies for the late response. I fell ill, and I'm not around the computer right now.
I will try to take a closer look at the changes you've made in a week or two; I will also take a look at CI/CD, to have Windows and ARM in the pipeline.
The UB-checker complaints that I mentioned are also present on the master branch, so it's something for me to check out.

**Review by aka-rider — APPROVED** [review]

**Pull request merged**